// neo-irc/neo-irc.js
//###################################################################################################

module.exports = {    
    //===============================================================================================
    
    //===============================================================================================
    // Name
    name: 'irc',

    //===============================================================================================
    // Depends
    depends: [],    
    
    //===============================================================================================
    // Config
    config: [
        'host',
        'name',
        'channel'
    ],
    
    //===============================================================================================
    // Properties
    properties: [
        "irc",
        "server"
    ],
    
    //===============================================================================================
    // Init
    init: function(ready) {        
        var self = this;

        this.irc = require('irc');
        this.server = new this.irc.Client(
            self.config.host, 
            self.config.name, 
            {channels: [self.config.channel]}
        );
        
        // Events
        this.server.addListener('message', function(from, to, message) {
            if( message !== undefined )
            {          
                // Remove IRC Coloring :X \x80-\xFF
                message = message.replace(/[\x00-\x1F]/g, "");
    
                if( message.charAt(0) === '!')
                    self.events.emit(self.name+'.command', from, to, message);
                else
                    self.events.emit(self.name+'.message', from, to, message);                
            }
        });
    
        this.server.addListener('names', function(channel, names) {
            self.events.emit(self.name+'.names', channel, names);
        });  
        
        this.server.addListener('topic', function(channel,topic, nick) {
            self.events.emit(self.name+'.topic', channel, topic, nick);
        });
    
        this.server.addListener('join', function(channel,nick) {
            self.events.emit(self.name+'.join', channel, nick);
        });
    
        this.server.addListener('part', function(channel,nick) {
            self.events.emit(self.name+'.part', channel, nick);
        });
        
        this.server.addListener('quit', function(nick, reason, channels) {
            self.events.emit(self.name+'.quit', channels, nick);
        });
        
        this.server.addListener('kick', function(channel,nick) {
            self.events.emit(self.name+'.kick', channel, nick);
        });
    
        this.server.addListener('nick', function(oldnick,newnick, channels) {
            self.events.emit(self.name+'.nick', oldnick, newnick, channels);
        });

        ready();
    },
    
    //===============================================================================================
    // Exit
    exit: function(ready) {
        var self = this;
        ready();
    },

    //===============================================================================================
    // Methods
    methods: [
    ],
    
    //===============================================================================================
    // Slots
    slots: [
        ////-----------------------------------------------------------------------------------------
        // Send an IRC Command
        function send() {     
            this.server.send.apply(this,arguments);
        },
        
        ////-----------------------------------------------------------------------------------------
        // Send an IRC message
        function say(to, message) {     
            this.server.say(to, message);               
        },
        
         ////-----------------------------------------------------------------------------------------
        // Send an IRC ACTION!
        function action(to, message) {     
            this.server.action(to, message);               
        },       
        
         ////-----------------------------------------------------------------------------------------
        // Sets the channel topic
        function settopic(channel, message) {     
            this.server.send('TOPIC', channel, message);               
        } 
    ],
    
    //===============================================================================================
    // Exports
    exports: [
    ]
    
    //===============================================================================================
};

//###################################################################################################
